**Week 8 Independent Lab**

# Advanced Data Manipulation

Author: Magdalena Gerharter
Course: BGEN 632 Grad. Intro. to Python
Term: Spring 2025

## Import modules

In [7]:
import os
import numpy as np
import pandas as pd

## Set working directory

In [10]:
os.chdir("/Users/magdalenagerharter/Documents/GitHub/week8labs/data")
os.getcwd()  #confirm change

* `blau_gender` change to `gender_blau`
* `Gini_gh_ten` change to `gh_tenure_gini`
* `Core1` change to `core_dev`
* `commits` change to `user_commits`
* `propCommits` change to `commit_proportion`
* `github_tenure` change to `gh_tenure`
* `leavesNextQ` change to `leaves_next_q`

'/Users/magdalenagerharter/Documents/GitHub/week8labs/data'

## Load data and Rename Columns

In [15]:
turnover = pd.read_csv("gh_turnover_four_projects.csv")

In [29]:
(turnover
 .rename({'domain':'project_domain',
          'language':'project_language',
          'windows':'project_age',
          'window_idx':'quarter',
          'num_team':'team_size',
          'num_commits':'project_commits',
          'blau_gender':'gender_blau',
          'Gini_gh_ten':'gh_tenure_gini',
          'Core1':'core_dev',
          'commits':'user_commits',
          'propCommits':'commit_proportion',
          'github_tenure':'gh_tenure',
          'leavesNextQ':'leaves_next_q'},
         axis = 1,
         inplace = True))
turnover

project_id project_domain project_language  project_age  quarter  \
0          1101          OTHER           Python           23       22   
1          1101          OTHER           Python           23       22   
2          1101          OTHER           Python           23       22   
3          1101          OTHER           Python           23       22   
4          1101          OTHER           Python           23       22   
..          ...            ...              ...          ...      ...   
468       24293      DEV_FRAME           Python           24       23   
469       24293      DEV_FRAME           Python           24       24   
470       24293      DEV_FRAME           Python           24       24   
471       24293      DEV_FRAME           Python           24       24   
472       24293      DEV_FRAME           Python           24       24   

     team_size  project_commits  gender_blau  gh_tenure_gini  user_id  \
0           26            22688        0.000        0.422259  1614129   
1           26            22688        0.000        0.422259  1614101   
2           26            22688        0.000        0.422259    94412   
3           26            22688        0.000        0.422259    88459   
4           26            22688        0.000        0.422259   738787   
..         ...              ...          ...             ...      ...   
468         12             1655        0.153        0.292578   938914   
469         12             1224        0.000        0.229926    11648   
470         12             1224        0.000        0.229926   179354   
471         12             1224        0.000        0.229926   438642   
472         12             1224        0.000        0.229926     5268   

     core_dev  user_commits  commit_proportion gender  gh_tenure  \
0           1         22138           0.975758   male        426   
1           0           124           0.005465   male         46   
2           0            53           0.002336   male        941   
3           0            49           0.002160   male       1870   
4           0            45           0.001983   male        781   
..        ...           ...                ...    ...        ...   
468         0             1           0.000604   male        326   
469         0             1           0.000817   male        893   
470         0             1           0.000817   male       1238   
471         0             1           0.000817   male       1582   
472         0             1           0.000817   male       1903   

     leaves_next_q  
0                0  
1                0  
2                1  
3                0  
4                0  
..             ...  
468              1  
469              0  
470              0  
471              0  
472              0  

[473 rows x 16 columns]

## Data Wrangling

### Mean of `user_commits`

In [35]:
turnover.user_commits.mean()

339.36575052854124

### Mean of `user_commits`

In [38]:
turnover.user_commits.median()

18.0

### Columns that start with 'p' or contain 'g' in `turnover_new`

In [53]:
turnover_new = turnover.filter(regex = "(^p)|(g)")
turnover_new

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
468       24293      DEV_FRAME           Python           24             1655   
469       24293      DEV_FRAME           Python           24             1224   
470       24293      DEV_FRAME           Python           24             1224   
471       24293      DEV_FRAME           Python           24             1224   
472       24293      DEV_FRAME           Python           24             1224   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.000        0.422259   male        426  
1          0.000        0.422259   male         46  
2          0.000        0.422259   male        941  
3          0.000        0.422259   male       1870  
4          0.000        0.422259   male        781  
..           ...             ...    ...        ...  
468        0.153        0.292578   male        326  
469        0.000        0.229926   male        893  
470        0.000        0.229926   male       1238  
471        0.000        0.229926   male       1582  
472        0.000        0.229926   male       1903  

[473 rows x 9 columns]

### `project_commits`> 9000 in `turnover_new`

In [55]:
turnover_new.query("project_commits > 9000")

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

### Combination of the last 2 steps

In [63]:
turnover_newer = (turnover
                  .filter(regex = "(^p)|(g)")
                  .query("project_commits > 9000"))
turnover_newer

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

### ends with `e` or contain `s`, and rows where `gh_tenure >= 1991`

In [66]:
(turnover
 .filter(regex="(e$)|(s)")
 .query("gh_tenure >= 1991"))

project_language  project_age  team_size  project_commits  user_id  \
339             Java           16          2                7    33272   
457           Python           23         20             3070    88459   

     user_commits  gh_tenure  leaves_next_q  
339             6       2006              0  
457             1       2054              0

### Mean and Median of `user_commits`

In [69]:
(turnover
 .filter(regex="(e$)|(s)|(^user_commits$)")
 .query("gh_tenure >= 1991")
 .user_commits
 .agg(["mean","median"]))

mean      3.5
median    3.5
Name: user_commits, dtype: float64

#### Analysing `user_commits` stats  

When analysing all users:  
- Mean: 339.36575052854124  
- Median: 18.0

When only analysing users with 'gh_tenure >= 1991' and from columns that end with either an 'e' or contain an 's':
- Mean: 3.5
- Median: 3.5

This comparison suggests that the filtered group (users who have been on GitHub longer) are less active when it comes to commits than the overall group. The overall dataset includes users with a large number of commits, which raises the mean.

## Additional Resources

Pandas - DataFrame Reference overview
https://www.w3schools.com/python/pandas/pandas_ref_dataframe.asp

I used it to look up/revise some commands